### PR TITLE
Tagtypes

### PR DESCRIPTION
**Pull request by jaminh**
The tagtypes command will allow filtering of the tagtypes listed in tagtype_list.py.

### CODE REVIEW

File: mopidy_mpd/protocol/connection.py [hunk 1]
```diff
@@ -51,3 +52,54 @@ def ping(context):
...
+        context.session.tagtypes.extend(enabled_types)
+    else:
+        raise exceptions.MpdArgError("Unknown sub command")
+
+    tagtypes = [
+        value
+        for value in context.session.tagtypes
+        if value in tagtype_list.TAGTYPE_LIST
+    ]
+    return [("tagtype", tagtype) for tagtype in tagtypes]
```
Comment (kingosticks) [inline mopidy_mpd/protocol/connection.py:105]: Does mpd print the current tagtypes for all subcommands or only when there's no subcommand?

File: mopidy_mpd/protocol/connection.py [hunk 2]
```diff
@@ -51,3 +52,54 @@ def ping(context):
...
+        Clear the list of tag types this client is interested in.
+
+        ``tagtypes all``
+
+        Announce that this client is interested in all tag types.
+    """
+    if not args:
+        pass
+    elif args[0] == "all":
+        context.session.tagtypes = tagtype_list.TAGTYPE_LIST[:]
```
Comment (kingosticks) [inline mopidy_mpd/protocol/connection.py:85]:
```suggestion
        context.session.tagtypes = tagtype_list.TAGTYPE_LIST
```

File: mopidy_mpd/protocol/connection.py [hunk 3]
```diff
@@ -51,3 +52,54 @@ def ping(context):
...
+
+        Clear the list of tag types this client is interested in.
+
+        ``tagtypes all``
+
+        Announce that this client is interested in all tag types.
+    """
+    if not args:
+        pass
+    elif args[0] == "all":
```
Comment (kingosticks) [inline mopidy_mpd/protocol/connection.py:84]:
```suggestion
    elif args[0].lower() == "all":
```
and for the others.

In fact, better doing something like the original version:
```
args = list(args)
    if args:
        subcommand = args.pop(0).lower()
```

File: mopidy_mpd/protocol/status.py [hunk 4]
```diff
@@ -41,10 +41,14 @@ def currentsong(context):
     """
     tl_track = context.core.playback.get_current_tl_track().get()
     stream_title = context.core.playback.get_stream_title().get()
+    print("Tagtypes: " + str(context.session.tagtypes))
```
Comment (kingosticks) [inline mopidy_mpd/protocol/status.py:44]: left over from debug?
```suggestion
```

File: mopidy_mpd/session.py [hunk 5]
```diff
@@ -21,6 +22,7 @@ def __init__(self, connection, config=None, core=None, uri_map=None):
         self.dispatcher = dispatcher.MpdDispatcher(
             session=self, config=config, core=core, uri_map=uri_map
         )
+        self.tagtypes = tagtype_list.TAGTYPE_LIST[:]
```
Comment (kingosticks) [inline mopidy_mpd/session.py:25]:
```suggestion
        self.tagtypes = tagtype_list.TAGTYPE_LIST
```
And this would benefit from being a set rather than a list.

File: mopidy_mpd/protocol/__init__.py [hunk 6]
```diff
@@ -20,8 +20,8 @@
 #: The MPD protocol uses ``\n`` as line terminator.
 LINE_TERMINATOR = b"\n"
 
-#: The MPD protocol version is 0.19.0.
-VERSION = "0.19.0"
+#: The MPD protocol version is 0.21.0.
+VERSION = "0.21.0"
```
Comment (girst) [inline mopidy_mpd/protocol/__init__.py:24]: we shouldn't advertise 0.21 support until we have implemented (at least partial) support for find/search [filters](https://mpd.readthedocs.io/en/latest/protocol.html#filters).

File: mopidy_mpd/translator.py [hunk 7]
```diff
@@ -18,7 +18,7 @@ def normalize_path(path, relative=False):
     return "/".join(parts)
 
 
-def track_to_mpd_format(track, position=None, stream_title=None):
+def track_to_mpd_format(track, position=None, stream_title=None, tagtypes=None):
```
Comment (kingosticks) [inline mopidy_mpd/translator.py:21]: Does `tagtypes` need to be optional?

File: mopidy_mpd/protocol/reflection.py [hunk 8]
```diff
@@ -81,25 +80,6 @@ def notcommands(context):
...
-    parameters = list(parameters)
-    if parameters:
-        subcommand = parameters.pop(0).lower()
-        if subcommand not in ("all", "clear", "disable", "enable"):
-            raise exceptions.MpdArgError("Unknown sub command")
-        # TODO: this is only a stub; we always return all tags anyhow.
-        return
-    return [("tagtype", tagtype) for tagtype in tagtype_list.TAGTYPE_LIST]
-
-
```
Comment (girst) [inline mopidy_mpd/protocol/reflection.py:102]: why did you move this from reflection to connection? (not per se against it, but an explanation would be nice)

File: mopidy_mpd/protocol/status.py (hunk 4 above)
Comment (girst) [inline mopidy_mpd/protocol/status.py:44]: stray debug-print?

File: mopidy_mpd/protocol/connection.py [hunk 9]
```diff
@@ -51,3 +52,54 @@ def ping(context):
...
+        context.session.tagtypes = tagtype_list.TAGTYPE_LIST[:]
+    elif args[0] == "clear":
+        context.session.tagtypes.clear()
+    elif args[0] == "disable":
+        context.session.tagtypes = [
+            value for value in context.session.tagtypes if value not in args[1:]
+        ]
+    elif args[0] == "enable":
+        enabled_types = [
+            value for value in args[1:] if value not in context.session.tagtypes
```
Comment (kingosticks) [inline mopidy_mpd/protocol/connection.py:94]: should filter out things that not valid tagtypes here else someone could fill this with junk

File: tests/protocol/test_connection.py [hunk 10]
```diff
@@ -26,3 +26,93 @@ def test_kill(self):
...
+        self.assertInResponse("tagtype: AlbumArtist")
+        self.assertInResponse("tagtype: AlbumArtistSort")
+        self.assertInResponse("tagtype: Title")
+        self.assertInResponse("tagtype: Track")
+        self.assertInResponse("tagtype: Name")
+        self.assertInResponse("tagtype: Genre")
+        self.assertInResponse("tagtype: Date")
+        self.assertInResponse("tagtype: Composer")
+        self.assertInResponse("tagtype: Performer")
+        self.assertInResponse("tagtype: Disc")
```
Comment (kingosticks) [inline tests/protocol/test_connection.py:44]: You added "Comment", do you want "Comment" here?

**Review by girst — CHANGES_REQUESTED** [review]
note: just an informal review, but thank you, very cool!

File: tests/protocol/test_connection.py [hunk 11]
```diff
@@ -26,3 +26,93 @@ def test_kill(self):
...
+        self.assertInResponse("tagtype: MUSICBRAINZ_ALBUMID")
+        self.assertInResponse("tagtype: MUSICBRAINZ_ALBUMARTISTID")
+        self.assertInResponse("tagtype: MUSICBRAINZ_TRACKID")
+        self.assertInResponse("OK")
+
+    def test_tagtypes_clear(self):
+        self.send_request("tagtypes clear")
+        self.assertEqualResponse("OK")
+
+    def test_tagtypes_all(self):
```
Comment (kingosticks) [inline tests/protocol/test_connection.py:55]: This test is only as good as the tagtypes you remember to check, and I think you are missing the new one you added: "Comment". Better to check there's a response line for everything in `TAGTYPE_LIST` ? Or count them, or something.

File: tests/protocol/test_connection.py [hunk 12]
```diff
@@ -26,3 +26,93 @@ def test_kill(self):
...
+        self.assertInResponse("tagtype: MUSICBRAINZ_TRACKID")
+        self.assertInResponse("OK")
+
+    def test_tagtypes_disable(self):
+        self.send_request("tagtypes all")
+        self.send_request(
+            "tagtypes disable MUSICBRAINZ_ARTISTID MUSICBRAINZ_ALBUMID "
+            "MUSICBRAINZ_ALBUMARTISTID MUSICBRAINZ_TRACKID"
+        )
+        self.assertInResponse("tagtype: Artist")
```
Comment (kingosticks) [inline tests/protocol/test_connection.py:82]: Needs `self.assertNotInResponse("tagtype: MUSICBRAINZ_ARTISTID")` etc else is doesnt actually test the disable function did anything.

File: mopidy_mpd/protocol/connection.py [hunk 13]
```diff
@@ -51,3 +52,54 @@ def ping(context):
...
+        Announce that this client is interested in all tag types.
+    """
+    if not args:
+        pass
+    elif args[0] == "all":
+        context.session.tagtypes = tagtype_list.TAGTYPE_LIST[:]
+    elif args[0] == "clear":
+        context.session.tagtypes.clear()
+    elif args[0] == "disable":
+        context.session.tagtypes = [
```
Comment (kingosticks) [inline mopidy_mpd/protocol/connection.py:89]: I imagine with enable/disable mpd will error if you don't specify at least one tag. Does this do nothing and return OK instead?

File: tests/protocol/test_connection.py [hunk 14]
```diff
@@ -26,3 +26,93 @@ def test_kill(self):
...
+
+    def test_tagtypes_enable(self):
+        self.send_request("tagtypes clear")
+        self.send_request("tagtypes enable Artist Album Title Track Name Genre")
+        self.assertInResponse("tagtype: Artist")
+        self.assertInResponse("tagtype: Album")
+        self.assertInResponse("tagtype: Title")
+        self.assertInResponse("tagtype: Track")
+        self.assertInResponse("tagtype: Name")
+        self.assertInResponse("tagtype: Genre")
```
Comment (kingosticks) [inline tests/protocol/test_connection.py:105]: Need to check at least one other tag is not in the response .

**Review by kingosticks — CHANGES_REQUESTED** [review]

File: mopidy_mpd/protocol/connection.py (hunk 1 above)
Comment (girst) [reply mopidy_mpd/protocol/connection.py:105]: Just tested 0.23.5 and it doesn't. I would therefore suggest returning early if no subcommand was given

File: mopidy_mpd/protocol/connection.py (hunk 2 above)
Comment (girst) [reply mopidy_mpd/protocol/connection.py:85]: i think `[:]` makes a copy of the list (we don't want to modify the list-of-all-tagtypes). as per my review, i'd suggest making ctx.s.tt a set instead, so all that can go away

Comment (kingosticks) [reply mopidy_mpd/protocol/connection.py:85]: I agree, a set is better.

File: mopidy_mpd/protocol/reflection.py (hunk 8 above)
Comment (jaminh) [reply mopidy_mpd/protocol/reflection.py:102]: The tagtypes command was listed under the connections settings section here: https://mpd.readthedocs.io/en/latest/protocol.html#connection. I moved it to be consistent with that.

File: mopidy_mpd/protocol/connection.py (hunk 1 above)
Comment (jaminh) [reply mopidy_mpd/protocol/connection.py]: I switched this back to the way the subcommand and parameters were being parsed originally.

File: mopidy_mpd/protocol/connection.py (hunk 2 above)
Comment (jaminh) [reply mopidy_mpd/protocol/connection.py]: Changed to a set.

File: mopidy_mpd/protocol/connection.py (hunk 13 above)
Comment (jaminh) [reply mopidy_mpd/protocol/connection.py]: After updates enable/disable without tags will do nothing and return OK.

File: tests/protocol/test_connection.py [hunk 15]
```diff
@@ -26,3 +26,93 @@ def test_kill(self):
...
+        self.assertInResponse("OK")
+
+    def test_tagtypes_disable_x(self):
+        self.send_request("tagtypes disable x")
```
[the remark this answers is not in the document]
Comment (jaminh) [reply tests/protocol/test_connection.py:132]: Invalid tags are now just ignored.

File: mopidy_mpd/protocol/connection.py [hunk 16]
```diff
@@ -51,3 +52,54 @@ def ping(context):
...
+            value for value in context.session.tagtypes if value not in args[1:]
+        ]
+    elif args[0] == "enable":
+        enabled_types = [
+            value for value in args[1:] if value not in context.session.tagtypes
+        ]
+        context.session.tagtypes.extend(enabled_types)
```
[the remark this answers is not in the document]
Comment (jaminh) [reply mopidy_mpd/protocol/connection.py]: Changed to set and reverted parsing code.

File: mopidy_mpd/protocol/connection.py [hunk 17]
```diff
@@ -51,3 +52,54 @@ def ping(context):
...
+    else:
+        raise exceptions.MpdArgError("Unknown sub command")
+
+    tagtypes = [
+        value
+        for value in context.session.tagtypes
+        if value in tagtype_list.TAGTYPE_LIST
```
[the remark this answers is not in the document]
Comment (jaminh) [reply mopidy_mpd/protocol/connection.py]: Invalid tag types are just ignored. The protocol docs didn't seem to specify what should happen for invalid tag types so hopefully that is ok?

File: tests/protocol/test_connection.py (hunk 10 above)
Comment (jaminh) [reply tests/protocol/test_connection.py:44]: I added comment because I saw that it was being used somewhere. I probably should add all of the tagtypes listed here https://mpd.readthedocs.io/en/latest/protocol.html#tags. There seem to be some additional tagtypes being used that are not in that list though (i.e. X-AlbumUri) so I'm not sure how to handle those.

File: mopidy_mpd/protocol/reflection.py (hunk 8 above)
Comment (girst) [reply mopidy_mpd/protocol/reflection.py:102]: oh, you're right!

File: mopidy_mpd/protocol/__init__.py [hunk 18]
```diff
@@ -20,7 +20,7 @@
 #: The MPD protocol uses ``\n`` as line terminator.
 LINE_TERMINATOR = b"\n"
 
-#: The MPD protocol version is 0.19.0.
+#: The MPD protocol version is 0.21.0.
```
Comment (girst) [inline mopidy_mpd/protocol/__init__.py:23]: comment doesn't match reality ;)

File: mopidy_mpd/protocol/connection.py [hunk 19]
```diff
@@ -51,3 +52,50 @@ def ping(context):
...
+            context.session.tagtypes.clear()
+        elif subcommand == "disable":
+            context.session.tagtypes.difference_update(parameters)
+        elif subcommand == "enable":
+            enabled_types = [
+                value
+                for value in parameters
+                if value in tagtype_list.TAGTYPE_LIST
+            ]
+            context.session.tagtypes.update(enabled_types)
```
Comment (girst) [inline mopidy_mpd/protocol/connection.py:99]: I think we should align our implementation with mpd, and raise an error for unknown tagtypes.

```suggestion
            if not all((value in tagtype_list.TAGTYPE_LIST for value in parameters)):
                raise exceptions.MpdArgError("Unknown tag type")
            context.session.tagtypes.update(parameters)
```

File: tests/protocol/test_connection.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (girst) [reply tests/protocol/test_connection.py:132]: see comment on connecton.py; i think we should raise an error (same as mpd)

File: mopidy_mpd/translator.py [hunk 20]
```diff
@@ -106,22 +106,36 @@ def track_to_mpd_format(track, position=None, stream_title=None):
...
+    if tagtypes is not None:
+        result = [
+            element for element in result if _has_value(tagtypes, *element)
+        ]
+    else:
+        result = [
+            element
+            for element in result
+            if _has_value(tagtype_list.TAGTYPE_LIST, *element)
+        ]
```
[the remark this answers is not in the document]
Comment (girst) [reply mopidy_mpd/translator.py:118]: resolved, but not changed?

**Review by girst — CHANGES_REQUESTED** [review]
Great work so far! This looks really nice, only a few issues left from my side :)

File: mopidy_mpd/protocol/connection.py (hunk 19 above)
Comment (jaminh) [reply mopidy_mpd/protocol/connection.py]: I changed it to raise an error for invalid tagtypes, but it will still just do nothing if no tagtypes are specified.

Comment (girst) [reply mopidy_mpd/protocol/connection.py]: oh, i didn't think about that! what do you think about adding this before the `if not all` check?

```
if not parameters:
    raise exceptions.MpdArgError("Not enough arguments")
```

Comment (girst) [reply mopidy_mpd/protocol/connection.py]: it would probably be best to put both checks into a small function to reduce duplication

File: mopidy_mpd/protocol/connection.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (kingosticks) [reply mopidy_mpd/protocol/connection.py]: Just to clarify as I think you already fixed this. Since the MPD protocol isn't thoroughly defined, in general we just try to do the same as the latest mpd does.

File: tests/protocol/test_connection.py (hunk 10 above)
Comment (kingosticks) [reply tests/protocol/test_connection.py:45]: I'm not sure about that. tagtypes used to be live in the "reflection" area of MPD protocol because clients are supposed to only use tags reported by that command. So we should probably only report tags we support. Listing the others doesn't have any benefit. Don't worry about our Mopidy-specific "X-something" tags, they are fine as they are but anything else we should probably stop including in our responses..

Comment (jaminh) [reply tests/protocol/test_connection.py:45]: Should the Mopidy-specific tags be added to the tagtypes list? What about other tags like "file", "Pos" (Position?), and "Id"? Should those be added to the tagtypes list as well, and therefore be something that clients can specify they are/aren't interested in, or should those still be treated separately? Also just to clarify I was going to change the tagtype list to only include the tags being used in the translator.py track_to_mpd_format function, was that what you meant by only including tags we actually support?

Comment (kingosticks) [reply tests/protocol/test_connection.py:45]: The tagtypes stuff is specifically related to track *metadata* fields. It shows clients what metadata tags the server understands so they avoid using unknown ones in their command arguments. It also clients to reduce response output they don't care about. Fields like `id` and `pos` are not metadata, they are essentially part of the MPD protocol and all servers will support them. Arguably a client also might not care about some of these but in general there's not much you can do without them so disabling them is a weird use-case. Plus I don't believe mpd supports disabling them so neither will we. So `id`, `pos` etc should continue not to be part of the tagtypes list.

In general, the Mopidy-specific `X-` tags are metadata so they should stay part of the tagtypes list (and thus included in the tests here). However, `X-AlbumImage` is no longer used anywhere in Mopidy-MPD so we could drop that one.

> I was going to change the tagtype list to only include the tags being used in the translator.py track_to_mpd_format function, was that what you meant by only including tags we actually support?

The tagtype list should be a superset of:
* all track metadata-like fields we output in the translator
* all tag fields we support passed in command arguments.

The latter means adding `comment` to the list is a good idea.

Comment (jaminh) [reply tests/protocol/test_connection.py:44]: Ok I think the tagtype list should now be reduced to only those actually used by mopidy-mpd.

File: tests/protocol/test_connection.py (hunk 11 above)
Comment (jaminh) [reply tests/protocol/test_connection.py:57]: I added a check to ensure the length of the response is the same as the tagtype list (plus 1 for the "OK" line).

File: README.rst [hunk 21]
```diff
@@ -140,7 +140,7 @@ This is a non-exhaustive list of MPD features that Mopidy doesn't support.
 - Replay gain is not supported.
 - ``stats`` does not provide any statistics.
 - ``decoders`` does not provide information about available decoders.
-- ``tagtypes`` is not supported.
+- ``tagtypes`` is not fully supported.
```
Comment (girst) [inline README.rst:143]: can this line be removed now? if not, what functionality is still missing? (i'm not saying you should implement that too, just want to know what this is about)

**Review by girst — APPROVED** [review]
Thanks a lot for your work! It'll still need @kingosticks' approval, but LGTM!

File: README.rst (hunk 21 above)
Comment (jaminh) [reply README.rst]: Good catch, when I first started working on this I had just stubbed out the subcommands when I changed that. As far as I know it should be fully supported now so I removed that line.

File: mopidy_mpd/protocol/tagtype_list.py [hunk 22]
```diff
@@ -3,19 +3,18 @@
     "ArtistSort",
     "Album",
     "AlbumArtist",
-    "AlbumArtistSort",
```
Comment (kingosticks) [inline mopidy_mpd/protocol/tagtype_list.py:6]: Don't think should remove this.

File: mopidy_mpd/session.py [hunk 23]
```diff
@@ -21,6 +22,7 @@ def __init__(self, connection, config=None, core=None, uri_map=None):
         self.dispatcher = dispatcher.MpdDispatcher(
             session=self, config=config, core=core, uri_map=uri_map
         )
+        self.tagtypes = set(tagtype_list.TAGTYPE_LIST)
```
Comment (kingosticks) [inline mopidy_mpd/session.py:25]: Better making `tagtype_list.TAGTYPE_LIST` itself a set.

File: mopidy_mpd/translator.py (hunk 7 above)
Comment (kingosticks) [reply mopidy_mpd/translator.py:21]: To be clearer, I don't think this new argument should be optional.

File: mopidy_mpd/translator.py [hunk 24]
```diff
@@ -106,22 +106,29 @@ def track_to_mpd_format(track, position=None, stream_title=None):
     if track.album and track.album.uri:
         result.append(("X-AlbumUri", track.album.uri))
 
-    result = [element for element in result if _has_value(*element)]
+    if tagtypes is None:
+        tagtypes = tagtype_list.TAGTYPE_LIST
```
Comment (kingosticks) [inline mopidy_mpd/translator.py:110]: Either make this a set or make `tagtype_list.TAGTYPE_LIST` itself a set.

File: mopidy_mpd/translator.py (hunk 7 above)
Comment (kingosticks) [reply mopidy_mpd/translator.py:21]: To be clear, my concern with this as optional was that if there was anywhere in this where we'd forgotten to supply `tagtypes`, that would be a bug.

File: mopidy_mpd/protocol/tagtype_list.py (hunk 22 above)
Comment (jaminh) [reply mopidy_mpd/protocol/tagtype_list.py]: It didn't appear this was actually being used anywhere, but I added it back in.

File: mopidy_mpd/translator.py (hunk 24 above)
Comment (jaminh) [reply mopidy_mpd/translator.py:110]: Changed tagtype_list.TAGTYPE_LIST to a set.

File: mopidy_mpd/session.py (hunk 23 above)
Comment (jaminh) [reply mopidy_mpd/session.py]: Done

**Review by kingosticks — APPROVED** [review]

**Comment by kingosticks** [comment]
Thank you (and @girst) for all your work on this.
